src/pulsar_producer.erl

```diff
@@ -61,106 +99,227 @@ callback_mode() -> [state_functions].
+                    producer_name :: atom(),
+                    opts :: map(),
+                    callback :: undefined | mfa() | fun((map()) -> ok),
+                    batch_size :: non_neg_integer(),
+                    requests :: #{sequence_id() => [{replayq:ack_ref(),
+                                                     [gen_statem:from()],
+                                                     [{timestamp(), [pulsar:message()]}]}]},
+                    last_bin :: binary()
+                   }.
+
+callback_mode() -> [state_functions, state_enter].
 
 start_link(PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts) ->
     gen_statem:start_link(?MODULE, [PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts], []).
 
-send(Pid, Message) ->
-    gen_statem:cast(Pid, {send, Message}).
-
-send_sync(Pid, Message) ->
-    send_sync(Pid, Message, 5000).
+-spec send(gen_statem:server_ref(), [pulsar:message()]) -> ok.
+send(Pid, Messages) ->
+    From = undefined,
+    erlang:send(Pid, ?SEND_REQ(From, Messages)),
+    ok.
 
-send_sync(Pid, Message, Timeout) ->
-    gen_statem:call(Pid, {send, Message}, Timeout).
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()]) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages) ->
+    send_sync(Pid, Messages, 5_000).
+
+-spec send_sync(gen_statem:server_ref(), [pulsar:message()], timeout()) ->
+          {ok, send_receipt()}
+        | {error, producer_connecting
+                | producer_disconnected
+                | term()}.
+send_sync(Pid, Messages, Timeout) ->
+    Caller = self(),
+    MRef = erlang:monitor(process, Pid),
+    erlang:send(Pid, ?SEND_REQ({Caller, MRef}, Messages)),
+    receive
+        {MRef, Response} ->
+            erlang:demonitor(MRef, [flush]),
+            Response;
+        {'DOWN', MRef, process, Pid, Reason} ->
+            error({producer_down, Reason})
+    after
+        Timeout ->
+            erlang:demonitor(MRef, [flush]),
+            receive
+                {MRef, Response} ->
+                    Response
+            after
+                0 ->
+                    error(timeout)
+            end
+    end.
 
 %%--------------------------------------------------------------------
-%% gen_server callback
+%% gen_statem callback
 %%--------------------------------------------------------------------
-init([PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts]) ->
+
+init([PartitionTopic, Server, ProxyToBrokerUrl, ProducerOpts0]) ->
+    process_flag(trap_exit, true),
     {Transport, BrokerServer} = pulsar_utils:parse_url(Server),
+    ProducerID = maps:get(producer_id, ProducerOpts0),
+    ReplayqCfg0 =
+        case maps:get(replayq_dir, ProducerOpts0, false) of
+            false ->
+                #{mem_only => true};
+            BaseDir ->
+                Dir = filename:join([BaseDir, PartitionTopic]),
```

Comment:
Is there a node-wide unique identifier for this client, if yes, we should consider using it in the path? 
Otherwise, when using this config in EMQX, we should make sure that `BaseDir` has:
1. Node name in the path, because in deployments like k8s, one data volume could be shared by different nodes
2. Rule/resource ID in the path, because there can be different data pipelines towards the same pulsar topic.

Reply:
:heavy_check_mark: 

When choosing the `replayq_dir` in the actions, we are using as the base dir:

- EMQX's data dir
- The node name
- The action ID

```erlang
filename:join([emqx:get_env(data_dir), replayq, node(), ActId])
```

Then, this information is going to be compounded with the partition topic in `pulsar_producer:init` as well.